src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,786 @@
+    NodeID shortest_path_via = SPECIAL_NODEID;
+    EdgeWeight shortest_path_weight = INVALID_EDGE_WEIGHT;
+
+    // The current via node during search spaces overlap stepping and an artificial
+    // weight (overlap factor * shortest path weight) we use as a termination criteria.
+    NodeID overlap_via = SPECIAL_NODEID;
+    EdgeWeight overlap_weight = INVALID_EDGE_WEIGHT;
+
+    // All via nodes in the overlapping search space (except the shortest path via node).
+    // Will be filtered and ranked and then used for s,via and via,t alternative paths.
+    std::vector<WeightedViaNode> candidate_vias;
+
+    // The logic below is a bit weird - here's why: we want to re-use the MLD routingStep for
+    // stepping our search space from s and from t. We don't know how far to overlap until we have
+    // the shortest path. Once we have the shortest path we can use its weight to terminate when
+    // we're over factor * weight. We have to set the weight for routingStep to INVALID_EDGE_WEIGHT
+    // so that stepping will continue even after we reached the shortest path upper bound.
+
+    const bool force_loop_forward = needsLoopForward(phantom_node_pair);
+    const bool force_loop_backward = needsLoopBackwards(phantom_node_pair);
+
+    EdgeWeight forward_heap_min = forward_heap.MinKey();
+    EdgeWeight reverse_heap_min = reverse_heap.MinKey();
+
+    while (forward_heap.Size() + reverse_heap.Size() > 0)
+    {
+        if (shortest_path_weight != INVALID_EDGE_WEIGHT)
+            overlap_weight = shortest_path_weight * kSearchSpaceOverlapFactor;
+
+        // Termination criteria - when we have a shortest path this will guarantee for our overlap.
+        const bool keep_going = forward_heap_min + reverse_heap_min < overlap_weight;
+
+        if (!keep_going)
+            break;
+
+        // Force forward step to not break early when we reached the middle, continue for overlap.
+        overlap_weight = INVALID_EDGE_WEIGHT;
+
+        if (!forward_heap.Empty())
+        {
+            routingStep<FORWARD_DIRECTION>(facade,
+                                           forward_heap,
+                                           reverse_heap,
+                                           overlap_via,
+                                           overlap_weight,
+                                           force_loop_forward,
+                                           force_loop_backward,
+                                           phantom_node_pair);
+
+            if (!forward_heap.Empty())
+                forward_heap_min = forward_heap.MinKey();
+
+            if (overlap_weight != INVALID_EDGE_WEIGHT && overlap_via != SPECIAL_NODEID)
+            {
+                if (shortest_path_via != SPECIAL_NODEID)
+                {
+                    candidate_vias.push_back(WeightedViaNode{overlap_via, overlap_weight});
+                }
+                else
+                {
+                    shortest_path_via = overlap_via;
```

Comment:
`shortest_path_via` and `shortest_path_weight` should stay in sync, right? Why do we keep track of them at all during the algorithm and not simply use the minimum-weight via candidate?